# BAMS 506 Project - Diet Planning Problem
`Student Name`:Haibin(Allen) Yu, Wendy(Mengyuan) Liu, Wenqiang(Roy) Yu, Sicheng(Sisly) Lyu \
`Student ID`:30979249, 86527371, 71090450, 72710643\
`Section`: BA 2 

## An Executive Summary
This report addresses the challenge of designing an optimized meal plan for a school lunch program that meets strict nutritional requirements while minimizing costs. The goal is to ensure that children receive a well-balanced and diverse diet, while adhering to budgetary constraints. The problem is explored through four key questions, each progressively improving upon the previous approach to account for practicality, variety, and long-term feasibility.

### `Part a & Part b `
In the initial optimization, we generated an optimal daily meal plan by minimizing the total cost while ensuring the necessary nutritional requirements were met. The optimal solution included a variety of food items, such as baked potatoes, spaghetti, oranges, milk, and pork, with precise servings calculated for each item to meet the dietary goals.

The total daily cost of this meal plan was **$2.43**, providing a cost-effective solution to meet nutritional needs. However, while the solution was mathematically optimal, it lacked diversity, leading us to further refine the model in later stages to enhance food variety.

### `Part c`
In the initial optimization, we generated an optimal daily meal plan by minimizing the total cost while ensuring the necessary nutritional requirements were met.

However, the results heavily focused on certain food groups, such as grains and fruits, with limited servings of vegetables, which are essential for a balanced diet. In response, we refined the model by categorizing foods into Vegetables, Proteins, Fruits, Grains/Pasta, and Dairy, ensuring that at least one serving from each food group was included in the daily plan.

Additionally, we **removed the 30% limit constraint that previously restricted any single food item from contributing more than 30% of the total calories and protein intake.** This change allowed the solver more flexibility in food selection, leading to a more cost-effective and nutritionally balanced solution.

As a result of this adjustment, the new meal plan provided more diversity and balance across food types while still minimizing costs. The total cost of this updated meal plan was **$2.55** per day, offering a more nutritionally balanced and feasible solution.
### `Part d`
We extended the optimization to a weekly meal plan, ensuring that the food selection varied across the week while still meeting daily nutritional needs. To achieve this, we introduced two key constraints:

1. Binary variables were used to prevent repetition, ensuring that at least one food item on each day was different from the previous day’s selection.
2. We limited each food type (excluding dairy) to appear no more than 4 times over the week, ensuring a broader distribution of food types across the entire week. The choice of 4 ensures no food type dominates, balancing variety while allowing flexibility.

The algebraic formulation shows how we minimized the total cost of the meal plan over the week while ensuring the minimum and maximum nutrient constraints were met daily. The model controlled food diversity by limiting consecutive day repetitions and managing the frequency of food types in the weekly plan.

The weekly meal plan generated includes a variety of food items such as lettuce, baked potatoes, apples, bananas, and milk, with daily costs ranging from **2.56 to 4.35 dollars** per day. The total weekly cost is **$23.28**, ensuring both affordability and diversity in the children's diet.

Shanshan: The ES part looks great.

______
## Introduction of Part a,b
We present an optimized solution to the school meal planning problem, where the goal is to provide nutritionally balanced and cost-efficient meals for school children. Using linear programming techniques, we developed a model that selects a combination of food items to meet daily nutritional requirements while minimizing the total cost per meal.

**The solution is specifically designed to address the following key objectives:**

- Ensure that meals provide the necessary intake of essential nutrients such as calories, protein, fat, and vitamins within prescribed limits.
- Minimize the daily cost of the meal plan, which is particularly critical given the limited budget available for the school’s food program.
- Ensure food variety by incorporating different food groups (vegetables, protein, grains, fruit, and dairy) to maintain a balanced diet.
- Prevent any single food item from contributing excessively to the overall nutrition (limiting single food items to 30% of total calories and protein intake).


### Part a 

***Explanation of Algebraic Formulation Elements***

$$
\begin{array}{rll}
x_i & = \text{number of servings of food item } i \text{ consumed per day} \\
c_i & = \text{cost of food item } i \text{ per serving} \\
N_{ni} & = \text{amount of nutrient } n \text{ in one serving of food item } i \\
\text{Min}_n & = \text{minimum daily requirement for nutrient } n \\
\text{Max}_n & = \text{maximum daily intake allowed for nutrient } n \\
N_{\text{calories}, i} & = \text{amount of calories in one serving of food item } i \\
T_{\text{calories}} & = \text{total daily calorie intake from all food items} \\
N_{\text{protein}, i} & = \text{amount of protein in one serving of food item } i \\
T_{\text{protein}} & = \text{total daily protein intake from all food items} \\
\end{array}
$$

***Objective***

The objective is to minimize the total cost of the food, subject to the following constraints:

- Calories intake between 1800 and 2400 kcal.
- Fat intake between 60 and 95 grams.
- Sodium intake between 1200 and 2200 mg.
- Carbohydrates, fiber, protein, vitamins, and minerals within the required range.
- No single food item should make up more than 30% of the total calories or protein intake
- Non negative.


***Algebraic formulation of the problem***

$$\begin{array}{rll}
 \text{minimize} & \sum_{i \in \text{Foods}} c_i \cdot x_i & \text{(Total cost of food)} \\
 \text{s.t.} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calories}} \cdot x_i \geq 1800 & \text{(Minimum calories)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calories}} \cdot x_i \leq 2400 & \text{(Maximum calories)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fat}} \cdot x_i \geq 60 & \text{(Minimum fat)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fat}} \cdot x_i \leq 95 & \text{(Maximum fat)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{sodium}} \cdot x_i \geq 1200 & \text{(Minimum sodium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{sodium}} \cdot x_i \leq 2200 & \text{(Maximum sodium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{carbs}} \cdot x_i \geq 240 & \text{(Minimum carbohydrates)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{carbs}} \cdot x_i \leq 400 & \text{(Maximum carbohydrates)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fiber}} \cdot x_i \geq 30 & \text{(Minimum fiber)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fiber}} \cdot x_i \leq 35 & \text{(Maximum fiber)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{protein}} \cdot x_i \geq 40 & \text{(Minimum protein)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{protein}} \cdot x_i \leq 55 & \text{(Maximum protein)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin A}} \cdot x_i \geq 2000 & \text{(Minimum vitamin A)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin A}} \cdot x_i \leq 6000 & \text{(Maximum vitamin A)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin C}} \cdot x_i \geq 45 & \text{(Minimum vitamin C)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin C}} \cdot x_i \leq 1200 & \text{(Maximum vitamin C)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calcium}} \cdot x_i \geq 1300 & \text{(Minimum calcium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calcium}} \cdot x_i \leq 3000 & \text{(Maximum calcium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{iron}} \cdot x_i \geq 8 & \text{(Minimum iron)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{iron}} \cdot x_i \leq 40 & \text{(Maximum iron)} \\
 & N_{i,\text{calories}} \cdot x_i \leq 0.30 \cdot \sum_{j \in \text{Foods}} N_{j,\text{calories}} \cdot x_j & \text{(No food supplies more than 30% of total calories)} \\
 & N_{i,\text{protein}} \cdot x_i \leq 0.30 \cdot \sum_{j \in \text{Foods}} N_{j,\text{protein}} \cdot x_j & \text{(No food supplies more than 30% of total protein)} \\
 & x_i \geq 0, & \forall i \in \text{Foods} \\
\end{array}
$$

### Part b

In [79]:
import pandas as pd
import gurobipy as gp
from gurobipy import GRB

In [80]:
file_path = 'food_data.xlsx'
xls = pd.ExcelFile(file_path)

sheet1 = pd.read_excel(xls, 'Sheet1')  # Nutrient limits
sheet2 = pd.read_excel(xls, 'Sheet2')  # Food items and nutritional info

print("Sheet1 Columns:", sheet1.columns.tolist())
print("Sheet2 Columns:", sheet2.columns.tolist())

Sheet1 Columns: ['Nutrient', 'Min', 'Max']
Sheet2 Columns: ['Food', 'Serving', 'Calories (kcal)', 'Fat (g)', 'Sodium (mg)', 'Carbs (g)', 'Fiber (g)', 'Protein (g)', 'Vitamin A (IU)', 'Vitamin C (mg)', 'Calcium (mg)', 'Iron (mg)', 'Cost ($/serving)']


In [81]:
nutrient_limits = {}
for index, row in sheet1.iterrows():
    nutrient_limits[row['Nutrient']] = (row['Min'], row['Max'])

In [82]:
foods = sheet2['Food']
calories = sheet2['Calories (kcal)']
fat = sheet2['Fat (g)']
sodium = sheet2['Sodium (mg)']
carbs = sheet2['Carbs (g)']
fiber = sheet2['Fiber (g)']
protein = sheet2['Protein (g)']
vitamin_a = sheet2['Vitamin A (IU)']
vitamin_c = sheet2['Vitamin C (mg)']
calcium = sheet2['Calcium (mg)']
iron = sheet2['Iron (mg)']
cost = sheet2['Cost ($/serving)']

In [83]:
model = gp.Model('diet')
servings = model.addVars(len(foods), vtype=GRB.CONTINUOUS, name="servings")
model.setObjective(gp.quicksum(servings[i] * cost[i] for i in range(len(foods))), GRB.MINIMIZE)

**Nutrient constraints**

In [85]:
model.addConstr(gp.quicksum(servings[i] * calories[i] for i in range(len(foods))) >= nutrient_limits['Calories (kcal)'][0], "cal_min")
model.addConstr(gp.quicksum(servings[i] * calories[i] for i in range(len(foods))) <= nutrient_limits['Calories (kcal)'][1], "cal_max")
model.addConstr(gp.quicksum(servings[i] * fat[i] for i in range(len(foods))) >= nutrient_limits['Fat (g)'][0], "fat_min")
model.addConstr(gp.quicksum(servings[i] * fat[i] for i in range(len(foods))) <= nutrient_limits['Fat (g)'][1], "fat_max")
model.addConstr(gp.quicksum(servings[i] * sodium[i] for i in range(len(foods))) >= nutrient_limits['Sodium (mg)'][0], "sodium_min")
model.addConstr(gp.quicksum(servings[i] * sodium[i] for i in range(len(foods))) <= nutrient_limits['Sodium (mg)'][1], "sodium_max")
model.addConstr(gp.quicksum(servings[i] * carbs[i] for i in range(len(foods))) >= nutrient_limits['Carbs (g)'][0], "carbs_min")
model.addConstr(gp.quicksum(servings[i] * carbs[i] for i in range(len(foods))) <= nutrient_limits['Carbs (g)'][1], "carbs_max")
model.addConstr(gp.quicksum(servings[i] * fiber[i] for i in range(len(foods))) >= nutrient_limits['Fiber (g)'][0], "fiber_min")
model.addConstr(gp.quicksum(servings[i] * fiber[i] for i in range(len(foods))) <= nutrient_limits['Fiber (g)'][1], "fiber_max")
model.addConstr(gp.quicksum(servings[i] * protein[i] for i in range(len(foods))) >= nutrient_limits['Protein (g)'][0], "protein_min")
model.addConstr(gp.quicksum(servings[i] * protein[i] for i in range(len(foods))) <= nutrient_limits['Protein (g)'][1], "protein_max")
model.addConstr(gp.quicksum(servings[i] * vitamin_a[i] for i in range(len(foods))) >= nutrient_limits['Vitamin A (IU)'][0], "vitamin_a_min")
model.addConstr(gp.quicksum(servings[i] * vitamin_a[i] for i in range(len(foods))) <= nutrient_limits['Vitamin A (IU)'][1], "vitamin_a_max")
model.addConstr(gp.quicksum(servings[i] * vitamin_c[i] for i in range(len(foods))) >= nutrient_limits['Vitamin C (mg)'][0], "vitamin_c_min")
model.addConstr(gp.quicksum(servings[i] * vitamin_c[i] for i in range(len(foods))) <= nutrient_limits['Vitamin C (mg)'][1], "vitamin_c_max")
model.addConstr(gp.quicksum(servings[i] * calcium[i] for i in range(len(foods))) >= nutrient_limits['Calcium (mg)'][0], "calcium_min")
model.addConstr(gp.quicksum(servings[i] * calcium[i] for i in range(len(foods))) <= nutrient_limits['Calcium (mg)'][1], "calcium_max")
model.addConstr(gp.quicksum(servings[i] * iron[i] for i in range(len(foods))) >= nutrient_limits['Iron (mg)'][0], "iron_min")
model.addConstr(gp.quicksum(servings[i] * iron[i] for i in range(len(foods))) <= nutrient_limits['Iron (mg)'][1], "iron_max")

<gurobi.Constr *Awaiting Model Update*>

**No single food should contribute more than 30% of total calories and protein**

In [87]:
total_calories = gp.quicksum(servings[i] * calories[i] for i in range(len(foods)))
total_protein = gp.quicksum(servings[i] * protein[i] for i in range(len(foods)))

for i in range(len(foods)):
    model.addConstr(servings[i] * calories[i] <= 0.30 * total_calories, f"cal_limit_{i}")
    model.addConstr(servings[i] * protein[i] <= 0.30 * total_protein, f"protein_limit_{i}")

**Print the results**

In [89]:
# Optimize the model
model.optimize()

if model.status == GRB.OPTIMAL:
    for i in range(len(foods)):
        if servings[i].x > 0:
            print(f"{foods[i]}: {servings[i].x:.2f} servings")
    print(f"Total Cost: ${model.objVal:.2f}")
else:
    print("No optimal solution found.")

Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.6.0 23G93)

CPU model: Apple M1 Max
Thread count: 10 physical cores, 10 logical processors, using up to 10 threads

Optimize a model with 80 rows, 30 columns and 2326 nonzeros
Model fingerprint: 0x7bdffd3a
Coefficient statistics:
  Matrix range     [6e-02, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [8e+00, 6e+03]
Presolve removed 10 rows and 0 columns
Presolve time: 0.00s
Presolved: 70 rows, 40 columns, 2073 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   2.294375e+02   0.000000e+00      0s
      20    2.4310277e+00   0.000000e+00   0.000000e+00      0s

Solved in 20 iterations and 0.01 seconds (0.00 work units)
Optimal objective  2.431027707e+00
Potatoes, Baked: 3.15 servings
Spaghetti W/ Sauce: 0.61 servings
Banana: 0.09 servings
Oranges: 4.54 servings
2% Lowfat Milk: 2.04 servings
Skim Milk: 1.00 servings
Wh

______
### Optimal Solution (Servings per Food)

- **Potatoes, Baked**: 3.15 servings
- **Spaghetti W/ Sauce**: 0.61 servings
- **Banana**: 0.09 servings
- **Oranges**: 4.54 servings
- **2% Lowfat Milk**: 2.04 servings
- **Skim Milk**: 1.00 servings
- **White Rice**: 0.16 servings
- **Pork**: 0.57 servings

### Total Cost

The total cost of the optimal meal plan is **$2.43** per day.

### Part c

## Introduction of Part c

In this optimization model, we are solving a diet planning problem where food items have been categorized into five types:

`Vegetables`, `Protein`, `Fruit`, `Grain/Pasta`, and `Dairy`.

The objective is to minimize the cost of the daily meal plan while ensuring that the nutritional requirements are met and that there is a minimum of one serving from each food category.

#### Initial Model (Part b):
In the original model from part b, the results were focused heavily on certain food groups, like grains and fruits, but lacked sufficient servings of vegetables. For instance, the recommendation included over 4 servings of oranges and several servings of potatoes, but no vegetables like broccoli or carrots. This does not align well with nutritional guidelines, which suggest a balance across all food groups, particularly with an emphasis on vegetables for essential vitamins and minerals.

Additionally, while the cost of the meal plan was optimized, the variety of food was limited, making the plan less appealing and potentially monotonous for children. Lack of diversity in food items could also lead to nutritional deficiencies over time.

#### Changes:
To make the recommendation more realistic and nutritionally balanced for a child’s daily intake, we made the following changes to the model:

1. **Inclusion of Variety Constraint**: To ensure a balanced diet, we introduced a new constraint that requires at least one serving from each of the major food categories (vegetables, protein, fruit, grain/pasta, and dairy). This guarantees that children receive a diverse range of nutrients essential for their health.
2. **Removal of the 30% Constraint**: The constraint that limited any single food item to contribute no more than 30% of the total calories or protein intake was removed. This allowed for greater flexibility in food selection, helping to reduce the total cost while maintaining nutritional balance.


#### Updated Model:
The updated model now accounts for food variety by ensuring that each food category is represented in the daily meal plan. This adjustment leads to a more diverse and nutritionally sound solution while still minimizing the cost. Additionally, by removing the 30% constraint, we allow the solver more flexibility to choose lower-cost food items without sacrificing the overall balance of the diet.

***Algebraic formulation(Part c)***
$$
\begin{array}{rll}
 \text{minimize} & \sum_{i \in \text{Foods}} c_i \cdot x_i & \text{(Total cost of food)} \\
 \text{s.t.} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calories}} \cdot x_i \geq 1800 & \text{(Minimum calories)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calories}} \cdot x_i \leq 2400 & \text{(Maximum calories)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fat}} \cdot x_i \geq 60 & \text{(Minimum fat)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fat}} \cdot x_i \leq 95 & \text{(Maximum fat)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{sodium}} \cdot x_i \geq 1200 & \text{(Minimum sodium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{sodium}} \cdot x_i \leq 2200 & \text{(Maximum sodium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{carbs}} \cdot x_i \geq 240 & \text{(Minimum carbohydrates)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{carbs}} \cdot x_i \leq 400 & \text{(Maximum carbohydrates)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fiber}} \cdot x_i \geq 30 & \text{(Minimum fiber)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{fiber}} \cdot x_i \leq 35 & \text{(Maximum fiber)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{protein}} \cdot x_i \geq 40 & \text{(Minimum protein)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{protein}} \cdot x_i \leq 55 & \text{(Maximum protein)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin A}} \cdot x_i \geq 2000 & \text{(Minimum vitamin A)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin A}} \cdot x_i \leq 6000 & \text{(Maximum vitamin A)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin C}} \cdot x_i \geq 45 & \text{(Minimum vitamin C)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{vitamin C}} \cdot x_i \leq 120 & \text{(Maximum vitamin C)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calcium}} \cdot x_i \geq 1300 & \text{(Minimum calcium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{calcium}} \cdot x_i \leq 3000 & \text{(Maximum calcium)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{iron}} \cdot x_i \geq 8 & \text{(Minimum iron)} \\
 & \sum_{i \in \text{Foods}} N_{i,\text{iron}} \cdot x_i \leq 40 & \text{(Maximum iron)} \\
 & \sum_{i \in \text{Foods}} x_i \geq 1, & \forall \, \text{Type} \in \text{(Vegetables, Protein, Fruit, Grain/Pasta, Dairy)} \, \text{(At least 1 serving per food type)} \\
 & x_i \geq 0, & \forall i \in \text{Foods} \\
\end{array}
$$


shanshan: I guess the constraint corresponding to "variety constraint" should be $\sum_{i\in Type}x_i\geq1$.

In [93]:
import pandas as pd
import gurobipy as gp
from gurobipy import GRB

# Load the Excel file data (replace with your file path if necessary)
file_path = 'food_data.xlsx'
xls = pd.ExcelFile(file_path)

# Read the necessary sheets
sheet1 = pd.read_excel(xls, 'Sheet1')  # Nutrient limits
sheet2 = pd.read_excel(xls, 'Sheet2')  # Food items and nutritional info

**Mapping Foods to Categories**

In [95]:
food_categories = {
    "Broccoli": "Vegetable",
    "Carrots, Raw": "Vegetable",
    "Corn": "Vegetable",
    "Lettuce, Iceberg,Raw": "Vegetable",
    "Peppers, Sweet, Raw": "Vegetable",
    "Tomato,Red,Ripe,Raw": "Vegetable",
    "Tofu": "Protein",
    "Roasted Chicken": "Protein",
    "Poached Eggs": "Protein",
    "Scrambled Eggs": "Protein",
    "Turkey": "Protein",
    "Beef": "Protein",
    "Pork": "Protein",
    "White Tuna in Water": "Protein",
    "Apple, Raw, w/Skin": "Fruit",
    "Banana": "Fruit",
    "Grapes": "Fruit",
    "Kiwifruit, Raw, Fresh": "Fruit",
    "Oranges": "Fruit",
    "Potatoes, Baked": "Grain/Pasta",
    "Bagels": "Grain/Pasta",
    "Wheat Bread": "Grain/Pasta",
    "White Bread": "Grain/Pasta",
    "Spaghetti W/ Sauce": "Grain/Pasta",
    "Oatmeal": "Grain/Pasta",
    "Couscous": "Grain/Pasta",
    "White Rice": "Grain/Pasta",
    "Macaroni, cooked": "Grain/Pasta",
    "2% Lowfat Milk": "Dairy",
    "Skim Milk": "Dairy"
}

# Apply the mapping to the 'Food' column to create the 'type' variable
sheet2['Type'] = sheet2['Food'].map(food_categories)

In [96]:
sheet2

,Food,Serving,Calories (kcal),Fat (g),Sodium (mg),Carbs (g),Fiber (g),Protein (g),Vitamin A (IU),Vitamin C (mg),Calcium (mg),Iron (mg),Cost ($/serving),Type
0,Broccoli,10 Oz Pkg,73.8,0.8,68.2,13.6,8.5,8.0,5867.4,160.2,159.0,2.3,0.16,Vegetable
1,"Carrots, Raw",1/2 Cup Shredded,23.7,0.1,19.2,5.6,1.6,0.6,15471.0,5.1,14.9,0.3,0.07,Vegetable
2,Corn,1/2 Cup,72.2,0.6,2.5,17.1,2.0,2.5,106.6,5.2,3.3,0.3,0.18,Vegetable
3,"Lettuce, Iceberg,Raw",1 Leaf,2.6,0.0,1.8,0.4,0.3,0.2,66.0,0.8,3.8,0.1,0.02,Vegetable
4,"Peppers, Sweet, Raw",1 Pepper,20.0,0.1,1.5,4.8,1.3,0.7,467.7,66.1,6.7,0.3,0.53,Vegetable
5,"Potatoes, Baked",1/2 Cup,171.5,0.2,15.2,39.9,3.2,3.7,0.0,15.6,22.7,4.3,0.06,Grain/Pasta
6,Tofu,1/4 block,88.2,5.5,8.1,2.2,1.4,9.4,98.6,0.1,121.8,6.2,0.31,Protein
7,Roasted Chicken,1 lb chicken,277.4,10.8,125.6,0.0,0.0,42.2,77.4,0.0,21.9,1.8,0.84,Protein
8,Spaghetti W/ Sauce,1 1/2 Cup,358.2,12.3,1237.1,58.3,11.6,8.2,3055.2,27.9,80.2,2.3,0.78,Grain/Pasta
9,"Tomato,Red,Ripe,Raw","1 Tomato, 2-3/5 In",25.8,0.4,11.1,5.7,1.4,1.0,766.3,23.5,6.2,0.6,0.27,Vegetable


In [97]:
foods = sheet2['Food']
calories = sheet2['Calories (kcal)']
fat = sheet2['Fat (g)']
sodium = sheet2['Sodium (mg)']
carbs = sheet2['Carbs (g)']
fiber = sheet2['Fiber (g)']
protein = sheet2['Protein (g)']
vitamin_a = sheet2['Vitamin A (IU)']
vitamin_c = sheet2['Vitamin C (mg)']
calcium = sheet2['Calcium (mg)']
iron = sheet2['Iron (mg)']
cost = sheet2['Cost ($/serving)']
types = sheet2['Type']  
# Initialize the Gurobi model
model = gp.Model('diet')

# Decision variables: servings of each food item
servings = model.addVars(len(foods), vtype=GRB.CONTINUOUS, name="servings")

# Objective: Minimize total cost
model.setObjective(gp.quicksum(servings[i] * cost[i] for i in range(len(foods))), GRB.MINIMIZE)

**Nutrient constraints**

In [99]:
model.setObjective(gp.quicksum(servings[i] * cost[i] for i in range(len(foods))), GRB.MINIMIZE)
model.addConstr(gp.quicksum(servings[i] * calories[i] for i in range(len(foods))) >= nutrient_limits['Calories (kcal)'][0], "cal_min")
model.addConstr(gp.quicksum(servings[i] * calories[i] for i in range(len(foods))) <= nutrient_limits['Calories (kcal)'][1], "cal_max")
model.addConstr(gp.quicksum(servings[i] * fat[i] for i in range(len(foods))) >= nutrient_limits['Fat (g)'][0], "fat_min")
model.addConstr(gp.quicksum(servings[i] * fat[i] for i in range(len(foods))) <= nutrient_limits['Fat (g)'][1], "fat_max")
model.addConstr(gp.quicksum(servings[i] * sodium[i] for i in range(len(foods))) >= nutrient_limits['Sodium (mg)'][0], "sodium_min")
model.addConstr(gp.quicksum(servings[i] * sodium[i] for i in range(len(foods))) <= nutrient_limits['Sodium (mg)'][1], "sodium_max")
model.addConstr(gp.quicksum(servings[i] * carbs[i] for i in range(len(foods))) >= nutrient_limits['Carbs (g)'][0], "carbs_min")
model.addConstr(gp.quicksum(servings[i] * carbs[i] for i in range(len(foods))) <= nutrient_limits['Carbs (g)'][1], "carbs_max")
model.addConstr(gp.quicksum(servings[i] * fiber[i] for i in range(len(foods))) >= nutrient_limits['Fiber (g)'][0], "fiber_min")
model.addConstr(gp.quicksum(servings[i] * fiber[i] for i in range(len(foods))) <= nutrient_limits['Fiber (g)'][1], "fiber_max")
model.addConstr(gp.quicksum(servings[i] * protein[i] for i in range(len(foods))) >= nutrient_limits['Protein (g)'][0], "protein_min")
model.addConstr(gp.quicksum(servings[i] * protein[i] for i in range(len(foods))) <= nutrient_limits['Protein (g)'][1], "protein_max")
model.addConstr(gp.quicksum(servings[i] * vitamin_a[i] for i in range(len(foods))) >= nutrient_limits['Vitamin A (IU)'][0], "vitamin_a_min")
model.addConstr(gp.quicksum(servings[i] * vitamin_a[i] for i in range(len(foods))) <= nutrient_limits['Vitamin A (IU)'][1], "vitamin_a_max")
model.addConstr(gp.quicksum(servings[i] * vitamin_c[i] for i in range(len(foods))) >= nutrient_limits['Vitamin C (mg)'][0], "vitamin_c_min")
model.addConstr(gp.quicksum(servings[i] * vitamin_c[i] for i in range(len(foods))) <= nutrient_limits['Vitamin C (mg)'][1], "vitamin_c_max")
model.addConstr(gp.quicksum(servings[i] * calcium[i] for i in range(len(foods))) >= nutrient_limits['Calcium (mg)'][0], "calcium_min")
model.addConstr(gp.quicksum(servings[i] * calcium[i] for i in range(len(foods))) <= nutrient_limits['Calcium (mg)'][1], "calcium_max")
model.addConstr(gp.quicksum(servings[i] * iron[i] for i in range(len(foods))) >= nutrient_limits['Iron (mg)'][0], "iron_min")
model.addConstr(gp.quicksum(servings[i] * iron[i] for i in range(len(foods))) <= nutrient_limits['Iron (mg)'][1], "iron_max")

<gurobi.Constr *Awaiting Model Update*>

**At least one serving comes from each "Type" category**

In [101]:
food_types = sheet2['Type'].unique()  # Get unique types

for food_type in food_types:
    type_indices = [i for i in range(len(foods)) if types[i] == food_type]
    model.addConstr(gp.quicksum(servings[i] for i in type_indices) >= 1, f"type_min_{food_type}")

# Optimize the model
model.optimize()

Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.6.0 23G93)

CPU model: Apple M1 Max
Thread count: 10 physical cores, 10 logical processors, using up to 10 threads

Optimize a model with 25 rows, 30 columns and 556 nonzeros
Model fingerprint: 0x9e2d6440
Coefficient statistics:
  Matrix range     [1e-01, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 6e+03]
Presolve removed 10 rows and 0 columns
Presolve time: 0.00s
Presolved: 15 rows, 40 columns, 303 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    0.0000000e+00   2.784375e+02   0.000000e+00      0s
      12    2.5513019e+00   0.000000e+00   0.000000e+00      0s

Solved in 12 iterations and 0.00 seconds (0.00 work units)
Optimal objective  2.551301858e+00


**Print the results**

In [103]:
if model.status == GRB.OPTIMAL:
    for i in range(len(foods)):
        if servings[i].x > 0:
            print(f"{foods[i]}: {servings[i].x:.2f} servings")
    print(f"Total Cost: ${model.objVal:.2f}")
else:
    print("No optimal solution found.")

Lettuce, Iceberg,Raw: 1.00 servings
Potatoes, Baked: 3.19 servings
Spaghetti W/ Sauce: 0.42 servings
Oranges: 5.43 servings
2% Lowfat Milk: 2.94 servings
Beef: 0.55 servings
White Rice: 0.12 servings
Pork: 0.45 servings
Total Cost: $2.55


__________
### Optimal Solution (Servings per Food)

- **Lettuce, Iceberg, Raw**: 1.00 servings
- **Potatoes, Baked**: 3.19 servings
- **Spaghetti W/ Sauce**: 0.42 servings
- **Oranges**: 5.43 servings
- **2% Lowfat Milk**: 2.94 servings
- **Skim Milk**: 0.78 servings
- **Beef**: 0.55 servings
- **White Rice**: 0.12 servings
- **Pork**: 0.45 servings

### Total Cost

The total cost of the meal plan is **$2.55** per day.


### Part d

## Introduction of Part d
Building on the model from Question c, where we removed the 30% limit on food item contributions, we aimed to create a more diverse weekly meal plan while continuing to meet nutritional requirements and minimize costs.

In Part d, we added two key constraints to ensure variety and limit repetition over the course of a week:

- Non-repetition of food items: We introduced a binary variable to ensure that each day’s meal plan includes at least one food item that is different from the previous day's meal. This constraint helps prevent consecutive day repetition of the same meal and promotes greater daily variety.
- Limiting weekly occurrences per food type: We implemented a constraint that ensures each food (e.g., Broccoli, Carrots) appears no more than 4 times per week, with the exception of dairy (only have 2 items). The limit of 4 was chosen as it represents a reasonable balance—ensuring that no food type dominates the week while still allowing flexibility in meal planning. The threshold of 4 also ensures that no single food is consumed more than half of the week, maintaining overall variety.
  
By introducing these constraints, our goal was to reduce food repetition across the week and ensure that different food types are evenly distributed, leading to a more balanced and diverse weekly meal plan.

shanshan: good adjustment!

### Algebraic formulation(Part d)

$$
\begin{array}{rll}
 \text{min} & Z = \sum_{d=1}^{7} \sum_{i=1}^{N} s_{d,i} \cdot \text{cost}_i \\
 \text{s.t.} & \sum_{i=1}^{N} s_{d,i} \cdot \text{calories}_i \geq \text{min\_calories}, \quad & \text{(Calories: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{calories}_i \leq \text{max\_calories}, \quad & \text{(Calories: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{fat}_i \geq \text{min\_fat}, \quad & \text{(Fat: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{fat}_i \leq \text{max\_fat}, \quad & \text{(Fat: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{sodium}_i \geq \text{min\_sodium}, \quad & \text{(Sodium: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{sodium}_i \leq \text{max\_sodium}, \quad & \text{(Sodium: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{carbs}_i \geq \text{min\_carbs}, \quad & \text{(Carbohydrates: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{carbs}_i \leq \text{max\_carbs}, \quad & \text{(Carbohydrates: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{fiber}_i \geq \text{min\_fiber}, \quad & \text{(Fiber: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{fiber}_i \leq \text{max\_fiber}, \quad & \text{(Fiber: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{protein}_i \geq \text{min\_protein}, \quad & \text{(Protein: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{protein}_i \leq \text{max\_protein}, \quad & \text{(Protein: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{vitamin\_a}_i \geq \text{min\_vitamin\_a}, \quad & \text{(Vitamin A: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{vitamin\_a}_i \leq \text{max\_vitamin\_a}, \quad & \text{(Vitamin A: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{vitamin\_c}_i \geq \text{min\_vitamin\_c}, \quad & \text{(Vitamin C: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{vitamin\_c}_i \leq \text{max\_vitamin\_c}, \quad & \text{(Vitamin C: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{calcium}_i \geq \text{min\_calcium}, \quad & \text{(Calcium: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{calcium}_i \leq \text{max\_calcium}, \quad & \text{(Calcium: max)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{iron}_i \geq \text{min\_iron}, \quad & \text{(Iron: min)} \\
 & \sum_{i=1}^{N} s_{d,i} \cdot \text{iron}_i \leq \text{max\_iron}, \quad & \text{(Iron: max)} \\
 & s_{d,i} \leq x_{d,i} \cdot M, \quad & \text{(Linking constraints)} \\
 & s_{d,i} \geq x_{d,i} \cdot m, \quad & \text{(where M is a large number, and m is a small positive number.)}  \\
 & \sum_{i=1}^{N} (x_{d,i} - x_{d-1,i}) \geq 1, \quad & \text{(Diversity constraints)} \\
 & \sum_{d=1}^{7} x_{d,i} \leq 4, \quad & \text{(Type diversity)} \\
 & \sum_{i \in T} s_{d,i} \geq 1 & \text{(Food type diversity)} \\
\end{array}
$$


shanshan: I understand that the diversity constraints should be $\sum^n_{i=1}|x_{d,i}-x_{d-1,i}|\geq 1$. To deal with this nonlinear constraints, we need to introduce additional auxiliary variables, which might be beyond the scope of BAMS506. Additionally, non-negative constraints are needed in this formulation.

In [135]:
import pandas as pd
import gurobipy as gp
from gurobipy import GRB

# Load the Excel file data (replace with your file path if necessary)
file_path = 'food_data.xlsx'
xls = pd.ExcelFile(file_path)

# Read the necessary sheets
sheet1 = pd.read_excel(xls, 'Sheet1')  # Nutrient limits
sheet2 = pd.read_excel(xls, 'Sheet2')  # Food items and nutritional info

nutrient_limits = {}
for index, row in sheet1.iterrows():
    nutrient_limits[row['Nutrient']] = (row['Min'], row['Max'])

food_categories = {
    "Broccoli": "Vegetable",
    "Carrots, Raw": "Vegetable",
    "Corn": "Vegetable",
    "Lettuce, Iceberg,Raw": "Vegetable",
    "Peppers, Sweet, Raw": "Vegetable",
    "Tomato,Red,Ripe,Raw": "Vegetable",
    "Tofu": "Protein",
    "Roasted Chicken": "Protein",
    "Poached Eggs": "Protein",
    "Scrambled Eggs": "Protein",
    "Turkey": "Protein",
    "Beef": "Protein",
    "Pork": "Protein",
    "White Tuna in Water": "Protein",
    "Apple, Raw, w/Skin": "Fruit",
    "Banana": "Fruit",
    "Grapes": "Fruit",
    "Kiwifruit, Raw, Fresh": "Fruit",
    "Oranges": "Fruit",
    "Potatoes, Baked": "Grain/Pasta",
    "Bagels": "Grain/Pasta",
    "Wheat Bread": "Grain/Pasta",
    "White Bread": "Grain/Pasta",
    "Spaghetti W/ Sauce": "Grain/Pasta",
    "Oatmeal": "Grain/Pasta",
    "Couscous": "Grain/Pasta",
    "White Rice": "Grain/Pasta",
    "Macaroni, cooked": "Grain/Pasta",
    "2% Lowfat Milk": "Dairy",
    "Skim Milk": "Dairy"
}

# Apply the mapping to the 'Food' column to create the 'type' variable
sheet2['Type'] = sheet2['Food'].map(food_categories)

foods = sheet2['Food']
calories = sheet2['Calories (kcal)']
fat = sheet2['Fat (g)']
sodium = sheet2['Sodium (mg)']
carbs = sheet2['Carbs (g)']
fiber = sheet2['Fiber (g)']
protein = sheet2['Protein (g)']
vitamin_a = sheet2['Vitamin A (IU)']
vitamin_c = sheet2['Vitamin C (mg)']
calcium = sheet2['Calcium (mg)']
iron = sheet2['Iron (mg)']
cost = sheet2['Cost ($/serving)']
types = sheet2['Type']  

# Initialize the Gurobi model
model = gp.Model('diet')

days = 7

# Decision variables: servings of each food item for each day
servings = model.addVars(days, len(foods), vtype=GRB.CONTINUOUS, name="servings")

# Binary variables to indicate if a food item is selected on a given day
selected = model.addVars(days, len(foods), vtype=GRB.BINARY, name="selected")

# Objective: Minimize total cost over all days
model.setObjective(gp.quicksum(servings[d, i] * cost[i] for d in range(days) for i in range(len(foods))), GRB.MINIMIZE)

# Nutrient constraints for each day
for d in range(days):
    model.addConstr(gp.quicksum(servings[d, i] * calories[i] for i in range(len(foods))) >= nutrient_limits['Calories (kcal)'][0], f"cal_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * calories[i] for i in range(len(foods))) <= nutrient_limits['Calories (kcal)'][1], f"cal_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * fat[i] for i in range(len(foods))) >= nutrient_limits['Fat (g)'][0], f"fat_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * fat[i] for i in range(len(foods))) <= nutrient_limits['Fat (g)'][1], f"fat_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * sodium[i] for i in range(len(foods))) >= nutrient_limits['Sodium (mg)'][0], f"sodium_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * sodium[i] for i in range(len(foods))) <= nutrient_limits['Sodium (mg)'][1], f"sodium_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * carbs[i] for i in range(len(foods))) >= nutrient_limits['Carbs (g)'][0], f"carbs_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * carbs[i] for i in range(len(foods))) <= nutrient_limits['Carbs (g)'][1], f"carbs_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * fiber[i] for i in range(len(foods))) >= nutrient_limits['Fiber (g)'][0], f"fiber_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * fiber[i] for i in range(len(foods))) <= nutrient_limits['Fiber (g)'][1], f"fiber_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * protein[i] for i in range(len(foods))) >= nutrient_limits['Protein (g)'][0], f"protein_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * protein[i] for i in range(len(foods))) <= nutrient_limits['Protein (g)'][1], f"protein_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * vitamin_a[i] for i in range(len(foods))) >= nutrient_limits['Vitamin A (IU)'][0], f"vitamin_a_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * vitamin_a[i] for i in range(len(foods))) <= nutrient_limits['Vitamin A (IU)'][1], f"vitamin_a_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * vitamin_c[i] for i in range(len(foods))) >= nutrient_limits['Vitamin C (mg)'][0], f"vitamin_c_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * vitamin_c[i] for i in range(len(foods))) <= nutrient_limits['Vitamin C (mg)'][1], f"vitamin_c_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * calcium[i] for i in range(len(foods))) >= nutrient_limits['Calcium (mg)'][0], f"calcium_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * calcium[i] for i in range(len(foods))) <= nutrient_limits['Calcium (mg)'][1], f"calcium_max_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * iron[i] for i in range(len(foods))) >= nutrient_limits['Iron (mg)'][0], f"iron_min_day_{d}")
    model.addConstr(gp.quicksum(servings[d, i] * iron[i] for i in range(len(foods))) <= nutrient_limits['Iron (mg)'][1], f"iron_max_day_{d}")

# Link servings and selected variables
for d in range(days):
    for i in range(len(foods)):
        model.addConstr(servings[d, i] <= selected[d, i] * 1000, f"link_servings_selected_day_{d}_food_{i}")
        model.addConstr(servings[d, i] >= selected[d, i] * 0.1, f"link_servings_selected_day_{d}_food_{i}_min")

# Ensure at least one food item is different each day
for d in range(1, days):
    model.addConstr(gp.quicksum(selected[d, i] - selected[d-1, i] for i in range(len(foods))) >= 1, f"different_food_day_{d}")

# Cumulative constraint to ensure diversity over the week
for i in range(len(foods)):
    if types[i] != "Dairy":
        model.addConstr(gp.quicksum(selected[d, i] for d in range(days)) <= 4, f"cumulative_food_{i}")

# Type diversity constraint
food_types = sheet2['Type'].unique()  # Get unique types

for d in range(days):
    for food_type in food_types:
        type_indices = [i for i in range(len(foods)) if types[i] == food_type]
        model.addConstr(gp.quicksum(servings[d, i] for i in type_indices) >= 1, f"type_diversity_{food_type}_day_{d}")

# Optimize the model
model.optimize()

if model.status == GRB.OPTIMAL:
    print("Optimal meal plan:")
    total_cost = 0
    for d in range(days):
        day_cost = sum(servings[d, i].x * cost[i] for i in range(len(foods)))
        total_cost += day_cost
        print(f"Day {d + 1} (Cost: ${day_cost:.2f}):")
        for i in range(len(foods)):
            if servings[d, i].x > 0:
                print(f"  {foods[i]}: {servings[d, i].x:.2f} servings")
    print(f"Total Cost: ${total_cost:.2f}")
else:
    print("No optimal solution found.")


Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (mac64[arm] - Darwin 23.6.0 23G93)

CPU model: Apple M1 Max
Thread count: 10 physical cores, 10 logical processors, using up to 10 threads

Optimize a model with 629 rows, 420 columns and 5288 nonzeros
Model fingerprint: 0x6effed6c
Variable types: 210 continuous, 210 integer (210 binary)
Coefficient statistics:
  Matrix range     [1e-01, 2e+04]
  Objective range  [2e-02, 8e-01]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+00, 6e+03]
Presolve removed 7 rows and 0 columns
Presolve time: 0.01s
Presolved: 622 rows, 420 columns, 5099 nonzeros
Variable types: 210 continuous, 210 integer (210 binary)

Root relaxation: objective 1.786127e+01, 465 iterations, 0.00 seconds (0.01 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0   17.86127    0   34          -   17.86127      -     -    0s
     0     0   17.9060

### Weekly Meal Plan

#### Day 1 Plan:
- **Lettuce, Iceberg, Raw**: 1.00 servings
- **Potatoes, Baked**: 1.28 servings
- **Apple, Raw, w/ Skin**: 8.16 servings
- **Grapes**: 1.93 servings
- **2% Lowfat Milk**: 3.89 servings
- **Beef**: 2.90 servings

**Total Cost**: $4.35

#### Day 2 Plan:
- **Lettuce, Iceberg, Raw**: 1.00 servings
- **Potatoes, Baked**: 1.00 servings
- **Apple, Raw, w/ Skin**: 7.44 servings
- **Banana**: 1.14 servings
- **Oranges**: 0.29 servings
- **2% Lowfat Milk**: 3.88 servings
- **Beef**: 2.91 servings

**Total Cost**: $3.76

#### Day 3 Plan:
- **Lettuce, Iceberg, Raw**: 1.00 servings
- **Potatoes, Baked**: 3.16 servings
- **Spaghetti W/ Sauce**: 0.52 servings
- **Oranges**: 5.18 servings
- **2% Lowfat Milk**: 2.93 servings
- **Turkey**: 0.50 servings
- **White Rice**: 0.11 servings
- **Pork**: 0.50 servings

**Total Cost**: $2.56

#### Day 4 Plan:
- **Carrots, Raw**: 0.10 servings
- **Lettuce, Iceberg, Raw**: 0.90 servings
- **Potatoes, Baked**: 0.90 servings
- **Apple, Raw, w/ Skin**: 7.34 servings
- **Banana**: 1.33 servings
- **Oranges**: 0.27 servings
- **White Bread**: 0.10 servings
- **2% Lowfat Milk**: 3.88 servings
- **Beef**: 2.90 servings

**Total Cost**: $3.76

#### Day 5 Plan:
- **Carrots, Raw**: 0.17 servings
- **Corn**: 0.13 servings
- **Tomato, Red, Ripe, Raw**: 0.70 servings
- **Apple, Raw, w/ Skin**: 2.34 servings
- **Oranges**: 5.05 servings
- **White Bread**: 6.33 servings
- **2% Lowfat Milk**: 2.66 servings
- **Turkey**: 0.29 servings
- **White Rice**: 0.10 servings
- **Pork**: 0.71 servings

**Total Cost**: $3.16

#### Day 6 Plan:
- **Carrots, Raw**: 0.13 servings
- **Corn**: 0.87 servings
- **Spaghetti W/ Sauce**: 0.45 servings
- **Banana**: 5.59 servings
- **Wheat Bread**: 0.10 servings
- **White Bread**: 0.10 servings
- **2% Lowfat Milk**: 3.85 servings
- **Skim Milk**: 0.10 servings
- **Turkey**: 0.41 servings
- **White Rice**: 0.56 servings
- **Pork**: 0.59 servings

**Total Cost**: $2.85

#### Day 7 Plan:
- **Carrots, Raw**: 0.14 servings
- **Corn**: 0.86 servings
- **Spaghetti W/ Sauce**: 0.38 servings
- **Banana**: 5.79 servings
- **Wheat Bread**: 0.10 servings
- **White Bread**: 0.10 servings
- **2% Lowfat Milk**: 3.89 servings
- **Skim Milk**: 0.10 servings
- **Turkey**: 0.10 servings
- **Beef**: 0.36 servings
- **White Rice**: 0.58 servings
- **Pork**: 0.54 servings

**Total Cost**: $2.84

### Total Weekly Cost: $23.28



________
## Summary
In the initial optimization **Part a/b**, we obtained an optimal solution that minimized the total cost of the daily meal plan while meeting the required nutritional constraints. However, the solution lacked human-centered considerations and did not provide enough variety in terms of food selection, which is essential for maintaining a balanced and enjoyable diet for children.

To address this, in **Part c**, we introduced a food categorization approach that ensured students received different types of food each day (e.g., vegetables, proteins, grains, dairy, and fruits). This adjustment aimed to improve the variety of food intake, ensuring that each meal plan was more diverse and nutritionally well-rounded. By incorporating this categorization, we enhanced the daily meal plan by not only meeting nutritional requirements but also ensuring the children consumed a balanced mix of food types.

In **Part d**, we extended this concept over a full week, aiming to ensure that food choices were diverse not only on a daily basis but also across the week. We introduced constraints that limited the repetition of the same food items across consecutive days, promoting a greater variety in food consumption throughout the week. This weekly optimization still satisfied the daily nutritional constraints while significantly reducing the repetition rate of foods, thus ensuring both nutritional balance and meal diversity on a weekly basis.

### Discussion:
While our model successfully generates a cost-efficient, nutritionally balanced and diverse meal plan that adheres to the given constraints, there are several limitations that could affect its practical application:

1. **Uniform Food Preferences Assumption**:
   The model assumes that all individuals have the same food preferences, which is not realistic in real-life scenarios. People from different regions or backgrounds may have distinct dietary habits, and some individuals might have allergies or intolerances to certain foods. These variations in food preferences and dietary restrictions are important considerations that the current model does not account for. If we had access to data on students’ food preferences, allergies, or regional preferences, we could further refine the model to account for these factors and improve the plan's suitability for a specific population.

2. **Fractional Servings**:
   The model output includes fractional servings for some food items, which might seem impractical when considering individual meal portions. However, in the context of a school feeding program where meals are prepared for hundreds or thousands of students, these fractional servings become less problematic. When scaling the meal plan for the entire school, the quantities would be multiplied by the total number of students, leading to manageable bulk quantities. Therefore, from a school-wide perspective, these fractional servings can still be efficiently implemented without causing logistical issues.
   
   
### Recommendations:
1. **Incorporate Food Preferences and Allergies**:
   To make the model more adaptable to real-life scenarios, we recommend incorporating data on students’ food preferences, cultural dietary habits, and allergies. This could be achieved by adding constraints that reflect these considerations, ensuring that the meal plans are not only nutritionally balanced and cost-effective but also aligned with the actual dietary needs and preferences of the students.

2. **Address Fractional Servings in Implementation**:
   While fractional servings may seem impractical on an individual level, they pose no significant issue when planning meals for large groups, such as a school population. It is recommended that the school uses the optimized daily meal plan and scales it up to the total number of students, which will eliminate the issue of fractional servings and allow for efficient food procurement and preparation.

By refining the model with these considerations, the "Feeding Hope" project can further improve its ability to provide nutritious, cost-effective meals that are tailored to the specific needs and preferences of the student population.

## ChatGPT Reference

**Our homework utilizes GPT solely for assistance in Gurobi coding and grammartical corrections. All mathematical formulations, constraints, objective functions, and problem design have been independently developed by ourselves.**

Shanshan: Good summary.
 
Overall: check++ 90